File: certior_guard/shell_parse.py
"""Map a shell command to capabilities against a shell-normalised form.

A raw ``\\bcurl\\b`` regex is trivially evaded — ``c""url``, ``FOO=1 curl``,
``/usr/bin/curl``, ``sudo curl``, ``C=curl; $C`` all run curl while dodging the
match. So the command is tokenised like a shell would (:mod:`shlex`), split into
simple commands, stripped of env-assignment prefixes and wrappers
(``sudo``/``env``/``timeout`` …), and ``VAR=val`` indirection is resolved; rules
run against that normal form.

Not a sandbox: command substitution is Turing-complete, so this raises the bar
from "fooled by quotes" to "needs real obfuscation" and flags dynamic constructs
(``$(…)``, ``eval``, ``base64 -d``) for conservative handling. Fail-safe: a parse
failure returns ``None`` so the caller treats the command as an opaque exec.
"""
from __future__ import annotations

import re
import shlex
from dataclasses import dataclass, field
from typing import List, Optional

from certior_guard.patterns import SECRET_INLINE, SECRET_PATH

# Commands that delegate to a command given in their arguments — peel them off
# so ``sudo curl`` / ``timeout 5 curl`` / ``xargs curl`` resolve to ``curl``.
_WRAPPERS = {
    "sudo", "doas", "command", "env", "nohup", "nice", "ionice", "stdbuf",
    "setsid", "time", "timeout", "watch", "xargs", "exec", "builtin",
    "caffeinate", "chrt",
}
# Tokens that separate simple commands. Redirections (``>`` ``>>`` ``<``) are
# handled separately so their targets can be inspected; ``|`` is also tracked
# as a pipe for the fetch-and-run-into-a-shell rule.
_OPERATORS = {"|", "||", "&&", ";", "&", "|&", "(", ")"}
_PIPES = {"|", "|&"}

_SHELLS = {"sh", "bash", "zsh", "dash", "ksh", "ash", "fish", "csh", "tcsh", "busybox"}
_DOWNLOADERS = {"curl", "wget", "fetch", "aria2c", "lynx", "w3m", "httpie", "http"}

_ASSIGN = re.compile(r"^[A-Za-z_]\w*=")
_VARREF = re.compile(r"^\$\{?([A-Za-z_]\w*)\}?$")
_DYNAMIC = re.compile(r"\$\(|`|\beval\b|\bsource\b|base64\s+-{0,2}d|\bxxd\b\s+-r|\bopenssl\b.*\benc\b.*-d")


@dataclass
class SimpleCommand:
    """One command in a pipeline/list, normalised to its bare executable."""
    executable: str          # basename, quotes removed, wrappers/assignments peeled
    args: List[str] = field(default_factory=list)

    @property
    def exe(self) -> str:
        return self.executable.lower()


@dataclass
class ParsedShell:
    commands: List[SimpleCommand]
    piped: bool              # the command line contains a pipe
    has_dynamic: bool        # command substitution / eval / base64-decode present
    redirect_targets: List[str] = field(default_factory=list)


def _basename(token: str) -> str:
    return token.rsplit("/", 1)[-1]


def _tokenize(command: str) -> Optional[List[str]]:
    """Shell-tokenise ``command`` (quote/escape removal, operators split out).

    Returns ``None`` on a parse error (e.g. unbalanced quotes) so the caller
    falls back to the legacy regex path rather than guessing.
    """
    # Treat newlines as command separators (shlex would swallow them as plain
    # whitespace and silently merge two commands into one).
    command = command.replace("\n", " ; ").replace("\r", " ")
    lex = shlex.shlex(command, posix=True, punctuation_chars=";&|<>()")
    lex.whitespace_split = True
    try:
        return list(lex)
    except ValueError:
        return None


def _substitute_vars(tokens: List[str]) -> List[str]:
    """Resolve simple ``VAR=val`` … ``$VAR`` indirection within one command line.

    Loose by design (ignores scope/ordering) — for a guard, resolving
    ``C=curl; $C`` to ``curl`` and erring toward caution is the right trade.
    """
    varmap = {}
    for t in tokens:
        if _ASSIGN.match(t):
            name, _, val = t.partition("=")
            varmap[name] = val
    if not varmap:
        return tokens
    out = []
    for t in tokens:
        m = _VARREF.match(t)
        if m and m.group(1) in varmap:
            out.append(varmap[m.group(1)])
        else:
            out.append(t)
    return out


def _peel(tokens: List[str]) -> Optional[SimpleCommand]:
    """Drop leading assignments and wrapper commands; return the real command."""
    i, n = 0, len(tokens)
    while i < n:
        t = tokens[i]
        if _ASSIGN.match(t):            # FOO=bar env-assignment prefix
            i += 1
            continue
        base = _basename(t)
        if base.lower() in _WRAPPERS:   # sudo / env / timeout 5 / xargs -I{} …
            i += 1
            while i < n and (tokens[i].startswith("-") or tokens[i].isdigit()):
                i += 1
            continue
        # Only the executable is reduced to a basename; arguments keep their
        # full text because paths matter (e.g. secret-file detection).
        return SimpleCommand(executable=base, args=list(tokens[i + 1:]))
    return None


def analyze(command: str) -> Optional[ParsedShell]:
    """Parse ``command`` into normalised simple commands, or ``None`` on failure."""
    if not command or not command.strip():
        return ParsedShell(commands=[], piped=False, has_dynamic=False)
    tokens = _tokenize(command)
    if tokens is None:
        return None
    tokens = _substitute_vars(tokens)

    piped = any(t in _PIPES for t in tokens)
    redirect_targets: List[str] = []
    commands: List[SimpleCommand] = []
    segment: List[str] = []
    expect_redirect_target = False
    for t in tokens:
        if expect_redirect_target:
            redirect_targets.append(t)
            expect_redirect_target = False
            continue
        if t in (">", ">>", "<", ">|"):
            expect_redirect_target = True
            continue
        if t in _OPERATORS:
            if segment:
                cmd = _peel(segment)
                if cmd:
                    commands.append(cmd)
                segment = []
            continue
        segment.append(t)
    if segment:
        cmd = _peel(segment)
        if cmd:
            commands.append(cmd)

    return ParsedShell(
        commands=commands,
        piped=piped,
        has_dynamic=bool(_DYNAMIC.search(command)),
        redirect_targets=redirect_targets,
    )


_READERS = {"cat", "less", "more", "head", "tail", "grep", "egrep", "strings",
            "xxd", "od", "base64", "bat", "nl", "tac", "awk", "sed", "cut"}
# Interpreters that run inline code (``python -c`` / ``node -e``) — a common way
# to read a secret file without naming a recognised reader command.
_INTERPRETERS = {"python", "python2", "python3", "node", "nodejs", "deno", "bun",
                 "ruby", "perl", "php", "rscript"}
# Commands that duplicate a file's contents; copying a secret out is a read.
_COPIERS = {"cp", "mv", "install", "ln"}
_DESTROYERS = {"dd", "shred", "wipefs", "blkdiscard"}
_DEPLOYERS = {"kubectl", "terraform", "helm", "serverless", "vercel", "netlify",
              "fly", "flyctl", "pulumi", "ansible-playbook", "aws"}
_DEPLOY_VERBS = {"apply", "deploy", "destroy", "delete", "up", "provision"}
# Package managers whose install/add pulls third-party code — a supply-chain surface.
_INSTALLERS = {"npm", "pnpm", "yarn", "pip", "pip3", "pipx", "uv", "poetry",
               "gem", "bundle", "cargo", "go", "apt", "apt-get", "brew", "apk"}
_INSTALL_VERBS = {"install", "add"}
_DB_DANGER = re.compile(r"\bdrop\b|\btruncate\b|\bdelete\s+from\b", re.IGNORECASE)


def _is_install(exe: str, args: List[str]) -> bool:
    if exe not in _INSTALLERS:
        return False
    if any(a in _INSTALL_VERBS for a in args):
        return True
    return (exe == "npm" and args[:1] == ["i"]) or (exe == "go" and "get" in args)


def shell_capabilities(command: str) -> Optional[List[str]]:
    """Map a Bash command to capability candidates via the parsed normal form.

    Returns a (possibly empty) list of capability strings, or ``None`` if the
    command could not be parsed — signalling the caller to fall back to the
    legacy regex rules. An empty list means "nothing dangerous matched".
    """
    parsed = analyze(command)
    if parsed is None:
        return None

    caps: List[str] = []

    def add(*cs: str) -> None:
        for c in cs:
            if c not in caps:
                caps.append(c)

    has_downloader = any(c.exe in _DOWNLOADERS for c in parsed.commands)
    has_shell = any(c.exe in _SHELLS for c in parsed.commands)

    # curl|wget … | sh  →  fetch-and-run remote code.
    if has_downloader and has_shell and parsed.piped:
        add("code:exec", "remote:exfiltrate")
    # bash -c "$(curl …)" / eval $(curl …) — dynamic fetch into a shell.
    if parsed.has_dynamic and (has_shell or any(c.exe == "eval" for c in parsed.commands)):
        if has_downloader or re.search(r"curl|wget", command, re.IGNORECASE):
            add("code:exec", "remote:exfiltrate")

    for c in parsed.commands:
        exe, args = c.exe, c.args

        if exe in _DESTROYERS or exe.startswith("mkfs"):
            add("fs:destroy")
        elif exe == "rm" and any(f.startswith("-") and ("r" in f or "f" in f) for f in args):
            add("fs:delete")
        elif exe == "git" and args[:1] == ["push"]:
            add("git:push")
        elif exe == "git" and args[:1] == ["merge"]:
            add("git:merge")
        elif exe in _DEPLOYERS and any(a in _DEPLOY_VERBS for a in args):
            add("prod:deploy")
        elif exe == "docker" and "push" in args:
            add("prod:deploy")
        elif exe in ("npm", "pnpm", "yarn") and "publish" in args:
            add("package:publish")
        elif exe == "twine" and "upload" in args:
            add("package:publish")
        elif exe == "cargo" and "publish" in args:
            add("package:publish")
        elif exe in ("psql", "mysql", "mongo", "mongosh") and _DB_DANGER.search(" ".join(args)):
            add("db:destroy")
        elif _is_install(exe, args):
            add("deps:install")
        elif exe in _READERS and any(SECRET_PATH.search(a) for a in args):
            add("secrets:read")
        elif exe in _INTERPRETERS and any(SECRET_INLINE.search(a) for a in args):
            # python -c "open('.env')" / node -e "readFileSync('.env')" / perl -pe 1 .env
            add("secrets:read")
        elif exe in ("source", ".") and any(SECRET_PATH.search(a) for a in args):
            # `source .env` / `. ./.env` — loads a secret file into the environment
            add("secrets:read")
        elif exe in _COPIERS and any(
            SECRET_PATH.search(a) for a in args[:-1] if not a.startswith("-")
        ):
            # `cp .env /tmp/x` — reads a secret out under a new (non-secret) name
            add("secrets:read")
        elif exe in ("scp", "rsync") and any("@" in a for a in args):
            add("data:exfiltrate")
        elif exe in _DOWNLOADERS and any(
            a in ("-F", "-d", "--data", "--data-binary", "-T", "--upload-file") for a in args
        ) and any(re.match(r"https?://(?!localhost|127\.)", a) for a in args):
            add("data:exfiltrate")

    # Writing to a raw block device wipes a disk regardless of the tool used.
    if any(re.match(r"/dev/(sd|nvme|mmcblk|vd|hd)", t) for t in parsed.redirect_targets):
        add("fs:destroy")

    return caps

File: src/shellParse.js
'use strict';
// Map a shell command to capabilities against a shell-normalised form. A raw
// `\bcurl\b` regex is trivially evaded — c""url, FOO=1 curl, /usr/bin/curl,
// sudo curl, C=curl; $C all run curl while dodging the match. So the command is
// tokenised like a shell would (quote/escape removal), split into simple
// commands, stripped of env-assignment prefixes and wrappers (sudo/env/timeout),
// and VAR=val indirection is resolved; rules run against that normal form.
//
// Not a sandbox: a parse failure returns null so the caller treats the command
// as an opaque exec, and dynamic constructs ($(…), eval, base64 -d) are flagged
// for conservative handling.
const { SECRET_PATH, SECRET_INLINE } = require('./patterns');

const WRAPPERS = new Set([
  'sudo', 'doas', 'command', 'env', 'nohup', 'nice', 'ionice', 'stdbuf',
  'setsid', 'time', 'timeout', 'watch', 'xargs', 'exec', 'builtin',
  'caffeinate', 'chrt',
]);
const OPERATORS = new Set(['|', '||', '&&', ';', '&', '|&', '(', ')']);
const PIPES = new Set(['|', '|&']);
const PUNCT = new Set([';', '&', '|', '<', '>', '(', ')']);

const SHELLS = new Set(['sh', 'bash', 'zsh', 'dash', 'ksh', 'ash', 'fish', 'csh', 'tcsh', 'busybox']);
const DOWNLOADERS = new Set(['curl', 'wget', 'fetch', 'aria2c', 'lynx', 'w3m', 'httpie', 'http']);

const ASSIGN = /^[A-Za-z_]\w*=/;
const VARREF = /^\$\{?([A-Za-z_]\w*)\}?$/;
const DYNAMIC = /\$\(|`|\beval\b|\bsource\b|base64\s+-{0,2}d|\bxxd\b\s+-r|\bopenssl\b.*\benc\b.*-d/;

function basename(token) { return token.slice(token.lastIndexOf('/') + 1); }

// Tokenise like a POSIX shell with punctuation split out. Returns null on a
// parse error (unbalanced quotes) so the caller falls back to opaque exec.
function tokenize(command) {
  command = command.replace(/\n/g, ' ; ').replace(/\r/g, ' ');
  const tokens = [];
  let i = 0;
  const n = command.length;
  while (i < n) {
    const c = command[i];
    if (c === ' ' || c === '\t') { i++; continue; }
    if (PUNCT.has(c)) {
      let j = i;
      while (j < n && PUNCT.has(command[j])) j++;
      tokens.push(command.slice(i, j));
      i = j;
      continue;
    }
    let word = '';
    while (i < n) {
      const ch = command[i];
      if (ch === ' ' || ch === '\t' || PUNCT.has(ch)) break;
      if (ch === '\\') { if (i + 1 < n) { word += command[i + 1]; i += 2; } else { i++; } continue; }
      if (ch === "'") {
        const end = command.indexOf("'", i + 1);
        if (end === -1) return null;
        word += command.slice(i + 1, end);
        i = end + 1;
        continue;
      }
      if (ch === '"') {
        let k = i + 1;
        let buf = '';
        let closed = false;
        while (k < n) {
          if (command[k] === '\\' && k + 1 < n && (command[k + 1] === '"' || command[k + 1] === '\\')) {
            buf += command[k + 1]; k += 2; continue;
          }
          if (command[k] === '"') { closed = true; break; }
          buf += command[k]; k++;
        }
        if (!closed) return null;
        word += buf;
        i = k + 1;
        continue;
      }
      word += ch; i++;
    }
    tokens.push(word);
  }
  return tokens;
}

// Resolve simple VAR=val … $VAR indirection within one command line.
function substituteVars(tokens) {
  const varmap = {};
  for (const t of tokens) {
    if (ASSIGN.test(t)) { const idx = t.indexOf('='); varmap[t.slice(0, idx)] = t.slice(idx + 1); }
  }
  if (Object.keys(varmap).length === 0) return tokens;
  return tokens.map((t) => {
    const m = VARREF.exec(t);
    return (m && m[1] in varmap) ? varmap[m[1]] : t;
  });
}

// Drop leading assignments and wrapper commands; return {exe, args} or null.
function peel(tokens) {
  let i = 0;
  const n = tokens.length;
  while (i < n) {
    const t = tokens[i];
    if (ASSIGN.test(t)) { i++; continue; }
    const base = basename(t);
    if (WRAPPERS.has(base.toLowerCase())) {
      i++;
      while (i < n && (tokens[i].startsWith('-') || /^\d+$/.test(tokens[i]))) i++;
      continue;
    }
    return { executable: base, exe: base.toLowerCase(), args: tokens.slice(i + 1) };
  }
  return null;
}

function analyze(command) {
  if (!command || !command.trim()) return { commands: [], piped: false, hasDynamic: false, redirects: [] };
  let tokens = tokenize(command);
  if (tokens === null) return null;
  tokens = substituteVars(tokens);

  const piped = tokens.some((t) => PIPES.has(t));
  const redirects = [];
  const commands = [];
  let segment = [];
  let expectTarget = false;
  for (const t of tokens) {
    if (expectTarget) { redirects.push(t); expectTarget = false; continue; }
    if (t === '>' || t === '>>' || t === '<' || t === '>|') { expectTarget = true; continue; }
    if (OPERATORS.has(t)) {
      if (segment.length) { const c = peel(segment); if (c) commands.push(c); segment = []; }
      continue;
    }
    segment.push(t);
  }
  if (segment.length) { const c = peel(segment); if (c) commands.push(c); }

  return { commands, piped, hasDynamic: DYNAMIC.test(command), redirects };
}

const READERS = new Set(['cat', 'less', 'more', 'head', 'tail', 'grep', 'egrep', 'strings',
  'xxd', 'od', 'base64', 'bat', 'nl', 'tac', 'awk', 'sed', 'cut']);
const INTERPRETERS = new Set(['python', 'python2', 'python3', 'node', 'nodejs', 'deno', 'bun',
  'ruby', 'perl', 'php', 'rscript']);
const COPIERS = new Set(['cp', 'mv', 'install', 'ln']);
const DESTROYERS = new Set(['dd', 'shred', 'wipefs', 'blkdiscard']);
const DEPLOYERS = new Set(['kubectl', 'terraform', 'helm', 'serverless', 'vercel', 'netlify',
  'fly', 'flyctl', 'pulumi', 'ansible-playbook', 'aws']);
const DEPLOY_VERBS = new Set(['apply', 'deploy', 'destroy', 'delete', 'up', 'provision']);
const INSTALLERS = new Set(['npm', 'pnpm', 'yarn', 'pip', 'pip3', 'pipx', 'uv', 'poetry',
  'gem', 'bundle', 'cargo', 'go', 'apt', 'apt-get', 'brew', 'apk']);
const INSTALL_VERBS = new Set(['install', 'add']);
const DB_DANGER = /\bdrop\b|\btruncate\b|\bdelete\s+from\b/i;

function isInstall(exe, args) {
  if (!INSTALLERS.has(exe)) return false;
  if (args.some((a) => INSTALL_VERBS.has(a))) return true;
  return (exe === 'npm' && args[0] === 'i') || (exe === 'go' && args.includes('get'));
}

// Returns a (possibly empty) array of capabilities, or null if unparseable.
function shellCapabilities(command) {
  const parsed = analyze(command);
  if (parsed === null) return null;

  const caps = [];
  const add = (...cs) => { for (const c of cs) if (!caps.includes(c)) caps.push(c); };

  const hasDownloader = parsed.commands.some((c) => DOWNLOADERS.has(c.exe));
  const hasShell = parsed.commands.some((c) => SHELLS.has(c.exe));

  if (hasDownloader && hasShell && parsed.piped) add('code:exec', 'remote:exfiltrate');
  if (parsed.hasDynamic && (hasShell || parsed.commands.some((c) => c.exe === 'eval'))) {
    if (hasDownloader || /curl|wget/i.test(command)) add('code:exec', 'remote:exfiltrate');
  }

  for (const c of parsed.commands) {
    const { exe, args } = c;
    if (DESTROYERS.has(exe) || exe.startsWith('mkfs')) add('fs:destroy');
    else if (exe === 'rm' && args.some((f) => f.startsWith('-') && (f.includes('r') || f.includes('f')))) add('fs:delete');
    else if (exe === 'git' && args[0] === 'push') add('git:push');
    else if (exe === 'git' && args[0] === 'merge') add('git:merge');
    else if (DEPLOYERS.has(exe) && args.some((a) => DEPLOY_VERBS.has(a))) add('prod:deploy');
    else if (exe === 'docker' && args.includes('push')) add('prod:deploy');
    else if (['npm', 'pnpm', 'yarn'].includes(exe) && args.includes('publish')) add('package:publish');
    else if (exe === 'twine' && args.includes('upload')) add('package:publish');
    else if (exe === 'cargo' && args.includes('publish')) add('package:publish');
    else if (exe === 'gh' && args.includes('release')) add('package:publish');
    else if (['psql', 'mysql', 'mongo', 'mongosh'].includes(exe) && DB_DANGER.test(args.join(' '))) add('db:destroy');
    else if (isInstall(exe, args)) add('deps:install');
    else if (READERS.has(exe) && args.some((a) => SECRET_PATH.test(a))) add('secrets:read');
    else if (INTERPRETERS.has(exe) && args.some((a) => SECRET_INLINE.test(a))) add('secrets:read');
    else if ((exe === 'source' || exe === '.') && args.some((a) => SECRET_PATH.test(a))) add('secrets:read');
    else if (COPIERS.has(exe)
      && args.slice(0, -1).some((a) => !a.startsWith('-') && SECRET_PATH.test(a))) add('secrets:read');
    else if (['scp', 'rsync'].includes(exe) && args.some((a) => a.includes('@'))) add('data:exfiltrate');
    else if (DOWNLOADERS.has(exe)
      && args.some((a) => ['-F', '-d', '--data', '--data-binary', '-T', '--upload-file'].includes(a))
      && args.some((a) => /^https?:\/\/(?!localhost|127\.)/.test(a))) add('data:exfiltrate');
  }

  if (parsed.redirects.some((t) => /^\/dev\/(sd|nvme|mmcblk|vd|hd)/.test(t))) add('fs:destroy');

  return caps;
}

module.exports = { analyze, shellCapabilities, tokenize };
